### DISCUSSION TITLE
something similar

### DISCUSSION THREAD

**Discussion by onyx-and-iris** [lebaston100/OBS-to-XAir#3] [Show and tell]
Hi! Thank you for putting together this obs-to-xair script. I'd like to share something similar I've written but it has some differences.

[obstoxair.py](https://gist.github.com/onyx-and-iris/a924fec392d690624bb4417984880327)

- supports websocket v5
- no support for x32 (the xair_api interface does not support it since the osc protocol for this mixer varies from the xair series)
- listens only for scene switch event.

It could be readily modified to support other obs events and other osc commands but I've tried to keep it similar to your code.

I could make a PR If you're interested just let me know. (perhaps on a second branch?)

Thanks for your time.

**Comment by lebaston100** [comment]
Hi!
If you want you can update the readme and create a PR and i'll merge it. Was mostly waiting for someone to request v5 support before putting in the work of getting and plugging in my mixer rack :D
Make sure to add credit to you in the readme if you do that.

**Comment by DJUnreal** [comment]
I'd really love to see X32 support re-added to this - the previous script worked perfectly with the X32 as long as you changed the port (it's off by one), as the mute state commands are the same in both, even though some of the other commands are not.

Additionally, I was using a modified version of this, which allowed multiple channel mute state changes for each scene change event. It's not my code, so I wasn't sure if I could/should share it, but essentially each switch had a list of channels to mute, a list to unmute, and anything not specifically configured was left unchanged. Is this something that could be added to this script too, for better functionality?

My use-case is that during live streams, I have different applications fed to different channels over the USB interface. Changing scenes, I want some of those sources untouched (eg my microphone for the most part), but some to be muted (eg Discord audio, game audio), and others to be unmuted (eg mini-game side audio). Changing back, it was unmuting the main game and Discord, muting the mini-game etc. This made it much easier for management of everything, as I could basically have a custom mix for each scene, rather than being locked to a single channel per scene.

I started with an XR18, but then replaced it with the X32 Rack when the XR18 I had failed and the store I bought it from said it was "beyond repair". The script migration was as simple as just changing the X-Air output port, and it "just worked" the same as previously.

**Comment by onyx-and-iris** [comment]
Hi!

Was it just mute states you were changing on scene changes? Also are you using websocket v5 with OBS?

It would be pretty simple to modify this script to mute/unmute sets of scenes, if lebaston likes that idea it could be investigated yes.

**Reply by DJUnreal** [reply]
Yes, I was just doing mute states, though given it uses OSC, I was considering adding more commands to it in the future. I never found a real use for that, though there were suggestions for ways it could be done in future if needed.

We basically changed the original script so that instead of reading for a single channel number, it read a list of them. Any channel numbers specified with a - after them got muted, and any specified without a -, or with a + after, got unmuted. Each scene change then triggered a list of mute/unmute commands, one for each specified channel.

It didn't add much - maybe an extra line or two - to the original script.

**Reply by lebaston100** [reply]
Feel free, i'm not gonna stop anyone making improvements on this.

**Comment by onyx-and-iris** [comment]
There are a couple of things to mention.

First:
This updated script uses the xair-api package and [here](https://github.com/onyx-and-iris/xair-api-python/blob/dev/xair_api/kinds.py) you'll noticed there is defined a kind map for the x32 but it would require some effort before it makes sense to add it to the interface. I don't have access to an x32 so I would rely on an x32 emulator for testing. I am willing to make this effort but you should give me a bit of time.

The interface does, however, have a send method in it's base class which can be used for sending osc commands directly like so:
`mixer.send("/ch/01/mix/on", 1)` , which would be an equivalent to `mixer.strip[0].mix.on = True` . This could be used for setting mute states for the x32 even before it is officially supported by the interface.

Also, the interface currently does support changing port with a keyword argument, for example:
`with xair_api.connect("XR18", port=10023)`

Second:
The updated script now uses obsws-python package which is written for OBS websocket v5. This is the new version of OBS websocket that is installed by default with OBS v28 +. I think it makes sense to support the new websocket version, many packages are rolling updates for it.

So you could probably make the changes yourself with the above information to get it working if you feel comfortable editing your code. Otherwise, if you're willing to share your code with me (you can send by email check my profile) I can get something working for now until I implement an update to officially support x32.

Thanks for the go-ahead lebaston.

**Reply by DJUnreal** [reply]
This makes perfect sense, thanks!

I'm happy to perform any testing you want done with my X32 here, if that helps? It's an X32 Rack, but they all talk the same X32 version of OSC (which I know is slightly different to the X-Air version, but a lot of it is common/shared). Shout if you want a hand with this, and I'm more than happy to pitch in. I'm not a developer myself, but have done a bit of basic debugging in the past.

I completely agree, supporting the new websocket version is absolutely the correct way to go about this. That's what broke the old script for me actually - the update to OBS Studio brought the new websocket with it, and that meant the script stopped working completely.

I think the old script did indeed use the `mixer.send("/ch/01/mix/on", 1)` style of commands, using the line `client.send_message("/ch/" + mychan + "/mix/on", mymute)` (mymute was part of the extra identifier we added, to determine whether we were muting a channel or not, by setting that to be 1 or 0 respectively depending on whether there was a trailing - on the channel number). It worked fine with the X32 Rack, just by changing the communication port.

Take as much time as you need - I've just restarted streaming after a couple of months of break (which is why I didn't spot the websocket change until last night, or I'd have asked for WS5 support sooner), and I've got a temporary workaround for the moment. It's not as elegant as this script though!

Thanks for even considering adding this!

**Reply by onyx-and-iris** [reply]
I would definitely appreciate that. I'll push x32 updates to a separate branch so you can check that anytime. I'll give you a shout when it's ready to test. Thanks for the offer!

**Reply by lebaston100** [reply]
Btw you can install obs-websocket 4.9 compat in the meantime to work with the old script https://github.com/obsproject/obs-websocket/releases/tag/4.9.1-compat

**Reply by DJUnreal** [reply]
Thanks both!

**Reply by onyx-and-iris** [reply]
Hi, 

I offer a proposal for a new version supporting X32. Setting the port should not be necessary, only setting the kind_mixer value (and of course conn info in `config.toml`)

Additionally, the scene to mute mapping should now be done in a mapping.toml file. 

If you want to run it through a test DJUnreal you may need to clone the fork and `pip install . -U` since you'll require latest version of the xair-api package.

I added an extra feature, ability to toggle groups. Not sure if it's worthwhile or necessary but the option is there.

If you like these changes/additions I shall make a pull request.

Let me know your thoughts lebaston.

[link to updated version](https://github.com/onyx-and-iris/OBS-to-XAir/tree/add_x32_support)

**Reply by lebaston100** [reply]
I like it a lot. What would you think about also exposing the mixer kind in the config.toml that way no .py editing would be necessary.

**Reply by onyx-and-iris** [reply]
Hey, thanks for the quick response. Yes that's a good suggestion although I'd like to keep that detail for the consumer library since the factory method generates an interface for all kinds and returns the one requested.

[proposed change](https://github.com/onyx-and-iris/OBS-to-XAir/blob/add_x32_support/__main__.py)

**Reply by lebaston100** [reply]
Looking good, awesome, feel free to pr when you feel it's ready and i'll merge.

**Reply by DJUnreal** [reply]
Wow, thanks guys! I was only out of the country on business for 2 days and I see you've done a fantastic job whilst I was away.

I'll give this a test tomorrow morning with my X32 Rack and confirm it all works properly; if anything doesn't, I'll put a separate issue report in for the specific whatever-doesn't-work part.

Thanks again to both of you for jumping on this so quickly and enthusiastically! You're both awesome!

**Reply by lebaston100** [reply]
it's all @onyx-and-iris at this point, i haven't done anything besides pushing a few project management buttons :D Big thanks to them.

**Reply by onyx-and-iris** [reply]
Hey,  @DJUnreal yes let me know if you run into any issues. 

Feel free to further investigate the xair-api as well. I have implemented the parts of the x32 api that overlap but there are also parts missing. If you run across something that seems like it really should be implemented let me know and I can investigate.

@lebaston100 thank you for the feedback, it was appreciated.

**Reply by DJUnreal** [reply]
I've had a chance to test it, and it works brilliantly! I do like the extra feedback you get in the script window - knowing it's detected a scene change is very handy indeed.

The mute and toggle functions are really nice also - it was good seeing toggle as an option in the display.

How is the mute/unmute translated? Does it only work for channels? I had an idea that it could be useful to have it mute MixBuses as well, because at the moment my first step when I start a stream is to mute the MixBus responsible for driving my speakers, so that I'm not blasting stream audio out of my local monitors as well as my earphones (I don't monitor my mic in the studio speakers, but do in the headphones). Again a nice-to-have, but if there's an easy way to address them it'd be potentially useful too.

I'm going to dig through XAir-API shortly, and see what I can see there too.

Massive thanks, again, for being so accommodating. Feel free to add to the readme that it's now been tested with an X32 Rack as well (it did indeed need me to use the -U flag to get it to detect, as I had the older version of XAir-API installed, and it didn't want to understand "X32" as a mixer type).

**Reply by onyx-and-iris** [reply]
Hey, thanks for the feedback.

mixer.strip refers to input channels, you can use mixer.bus for output buses.

Higher classes and methods are detailed in the README of the xair-api. Although it was written to cover XAir a lot will apply to the X32.
